Test suite fails with traitlets 5.8 and ipywidgets 8.0.4

```python
[   45s] ____________________ test_dataunion_constricts_widget_data _____________________
[   45s] 
[   45s]     def test_dataunion_constricts_widget_data():
[   45s]         class Foo(HasTraits):
[   45s]             bar = DataUnion(shape_constraint=shape_constraints(None, None, 3))
[   45s]     
[   45s]         ok_data = np.ones((4, 2, 3))
[   45s]         bad_data = np.ones((4, 4))
[   45s]         w = NDArrayWidget(ok_data)
[   45s]         foo = Foo(bar=w)
[   45s]         with pytest.raises(TraitError):
[   45s] >           w.array = bad_data
[   45s] E           Failed: DID NOT RAISE <class 'traitlets.traitlets.TraitError'>
[   45s] 
[   45s] ../../BUILDROOT/python-ipydatawidgets-4.3.2-32.154.x86_64/usr/lib/python3.8/site-packages/ipydatawidgets/tests/test_union.py:92: Failed
[   45s] ____________________ test_dataunion_widget_change_notified _____________________
[   45s] 
[   45s] mock_comm = <ipydatawidgets.tests.conftest.DummyComm object at 0x7f6c9f1c8d30>
[   45s] 
[   45s]     def test_dataunion_widget_change_notified(mock_comm):
[   45s]         ns = {'counter': 0}
[   45s]         class Foo(Widget):
[   45s]             bar = DataUnion().tag(sync=True)
[   45s]     
[   45s]             @observe('bar')
[   45s]             def on_bar_change(self, change):
[   45s]                 ns['counter'] += 1
[   45s]     
[   45s]         raw_data = np.ones((4, 4))
[   45s]         raw_data2 = np.ones((4, 4, 2))
[   45s]         w = NDArrayWidget(raw_data)
[   45s]     
[   45s]         foo = Foo(bar=w)
[   45s]         foo.comm = mock_comm
[   45s]         assert ns['counter'] == 1
[   45s]         w.array = raw_data2
[   45s] >       assert ns['counter'] == 2
[   45s] E       assert 1 == 2
[   45s] 
[   45s] ../../BUILDROOT/python-ipydatawidgets-4.3.2-32.154.x86_64/usr/lib/python3.8/site-packages/ipydatawidgets/tests/test_union.py:114: AssertionError
[   45s] _____________________ test_datawidget_creation_blank_comm ______________________
[   45s] 
[   45s] mock_comm = <ipydatawidgets.tests.conftest.DummyComm object at 0x7f6c9f1ab9d0>
[   45s] 
[   45s]     def test_datawidget_creation_blank_comm(mock_comm):
[   45s]         # The mock comm delays serialization, so error is avoided
[   45s]         # (the array serialization of undefined cause error)
[   45s] >       w = NDArrayWidget(comm=mock_comm)
[   45s] 
[   45s] ../../BUILDROOT/python-ipydatawidgets-4.3.2-32.154.x86_64/usr/lib/python3.8/site-packages/ipydatawidgets/tests/test_widgets.py:26: 
[   45s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[   45s] ../../BUILDROOT/python-ipydatawidgets-4.3.2-32.154.x86_64/usr/lib/python3.8/site-packages/ipydatawidgets/ndarray/widgets.py:70: in __init__
[   45s]     super(NDArrayWidget, self).__init__(array=array, **kwargs)
[   45s] /usr/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:502: in __init__
[   45s]     super().__init__(**kwargs)
[   45s] /usr/lib/python3.8/site-packages/traitlets/traitlets.py:1367: in __init__
[   45s]     self.notify_change(changes[key])
[   45s] /usr/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:693: in notify_change
[   45s]     self.send_state(key=name)
[   45s] /usr/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:571: in send_state
[   45s]     state = self.get_state(key=key)
[   45s] /usr/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:608: in get_state
[   45s]     value = to_json(getattr(self, k), self)
[   45s] ../../BUILDROOT/python-ipydatawidgets-4.3.2-32.154.x86_64/usr/lib/python3.8/site-packages/ipydatawidgets/ndarray/serializers.py:87: in array_to_compressed_json
[   45s]     state = array_to_json(value, widget)
[   45s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[   45s] 
[   45s] value = traitlets.Undefined, widget = NDArrayWidget()
[   45s] 
[   45s]     def array_to_json(value, widget):
[   45s]         """Array JSON serializer."""
[   45s]         if value is None:
[   45s]             return None
[   45s]         if value is Undefined:
[   45s] >           raise TraitError('Cannot serialize undefined array!')
[   45s] E           traitlets.traitlets.TraitError: Cannot serialize undefined array!
[   45s] 
[   45s] ../../BUILDROOT/python-ipydatawidgets-4.3.2-32.154.x86_64/usr/lib/python3.8/site-packages/ipydatawidgets/ndarray/serializers.py:56: TraitError
[   45s] =============================== warnings summary ===============================
[   45s] ../../../../../usr/lib/python3.8/site-packages/traittypes/traittypes.py:20
[   45s]   /usr/lib/python3.8/site-packages/traittypes/traittypes.py:20: DeprecationWarning: 
[   45s]               Sentinel is not a public part of the traitlets API.
[   45s]               It was published by mistake, and may be removed in the future.
[   45s]               
[   45s]     Empty = Sentinel('Empty', 'traittypes',
[   45s] 
[   45s] tests/test_ndarray_trait.py: 11 warnings
[   45s]   /usr/lib/python3.8/site-packages/traitlets/traitlets.py:703: DeprecationWarning: Explicit using of Undefined as the default value is deprecated in traitlets 5.0, and may cause exceptions in the future.
[   45s]     return self.get(obj, cls)
[   45s] 
[   45s] -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
[   45s] =========================== short test summary info ============================
[   45s] FAILED tests/test_union.py::test_dataunion_constricts_widget_data - Failed: D...
[   45s] FAILED tests/test_union.py::test_dataunion_widget_change_notified - assert 1 ...
[   45s] FAILED tests/test_widgets.py::test_datawidget_creation_blank_comm - traitlets...
[   45s] ================== 3 failed, 64 passed, 12 warnings in 0.85s ===================
```

--- a/ipydatawidgets/ndarray/union.py
+++ b/ipydatawidgets/ndarray/union.py
@@ -55,6 +55,10 @@ class DataUnion(Union):
             # comparison above returns something other than True/False
             obj._notify_trait(self.name, old_value, new_value)
 
+    def subclass_init(self, cls):
+        cls._instance_inits.append(self.instance_init)
+        return super().subclass_init(cls)
+
     def instance_init(self, inst):
         inst.observe(self._on_instance_value_change, self.name)
 

--- a/ipydatawidgets/tests/test_widgets.py
+++ b/ipydatawidgets/tests/test_widgets.py
@@ -24,8 +24,12 @@ def test_datawidget_creation_blank():
 def test_datawidget_creation_blank_comm(mock_comm):
     # The mock comm delays serialization, so error is avoided
     # (the array serialization of undefined cause error)
-    w = NDArrayWidget(comm=mock_comm)
-    assert w.array is Undefined
+    try:
+        w = NDArrayWidget(comm=mock_comm)
+    except TraitError as e:
+        assert 'Cannot serialize undefined array' in str(e)
+    else:
+        assert w.array is Undefined
 
 
 def test_datawidget_creation():

